enhancements/ocp-coreos-layering.md

```diff
@@ -0,0 +1,496 @@
+
+### Risks and Mitigations
+
+We're introducing a whole new level of customization for nodes, and because this functionality will be new, we don't yet have significant experience with it.  There are likely a number of potentially problematic "unknown unknowns".
+
+To say this another way: until now we've mostly stuck to the model that user code should run in a container, and keep the host relatively small.  This could be perceived as a major backtracking on that model.
+
+This also intersects heavily with things like [out of tree drivers](https://github.com/openshift/enhancements/pull/357).
+
+We will need some time to gain experience with what works and best practices, and develop tooling and documentation.
+
+It is likely that the initial version will be classified as "Tech Preview" from the OCP product perspective.
+
+#### Supportability of two update mechanisms
+
+If for some reason we cannot easily upgrade existing FCOS/RHCOS systems provisioned prior to the existence of this functionality, and hence need to support *two* ways to update CoreOS nodes, it will become an enormous burden.
+
+Also relatedly, we would need to continue to support [openshift-ansible](https://github.com/openshift/openshift-ansible) for some time alongside the `once-from` functionality.  See also [this issue](https://github.com/openshift/machine-config-operator/issues/1592).
+
+#### Versioning of e.g. kubelet
+
+We will need to ensure that we detect and handle the case where core components e.g. the `kubelet` binary is coming from the wrong place, or is the wrong version.
+
+
+#### Location of builds
+
+Today, ideally nodes are isolated from each other.  A compromised node can in theory only affect pods which land on that node.
+In particular we want to avoid a compromised worker node being able to easily escalate  compromise the control plane.
+
+#### Registry availability
+
+If implemented in the obvious way, OS updates would fail if the cluster-internal registry is down.
+
+A strong mitigation for this is to use ostree's native ability to "stage" the update across all machines before starting any drain at all.  However, we should probably still be careful to only stage the update on one node at a time (or `maxUnavailable`) in order to avoid "thundering herd" problems, particularly for the control plane with etcd.
+
+Another mitigation here may be to support peer-to-peer upgrades, or have the control plane host a "bootstrap registry" that just contains the pending OS update.
+
+#### Manifest list support
+
+We know we want heterogeneous clusters, right now that's not supported by the build and image stream APIs.
+
+
+#### openshift-install bootstrap node process
+
+A key question here is whether we need the OpenShift build API as part of the bootstrap node or not.  One option is to do a `podman build` on the bootstrap node.
```

Comment:
> (reimplementing a lot of push/pull secrets, proxy configs, CAs, ICSP logic)

We already need to be able to pull container images on the bootstrap node with plain `podman`, and we already do.  So that covers CAs and ICSP too.

The push aspect is interesting, but I think the most likely thing here is to run a basic in-memory registry (possibly the in-cluster one, just in a non-production mode) and we can easily set things up to allow pushing to that.  Or we just use `oc image serve` etc.